date-time/src/commonMain/kotlin/com/splendo/kaluga/datetime/timer/RecurringTimer.kt

```diff
@@ -0,0 +1,190 @@
+/*
+ Copyright 2021 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+package com.splendo.kaluga.datetime.timer
+
+import com.splendo.kaluga.state.HandleAfterNewStateIsSet
+import com.splendo.kaluga.state.HandleBeforeOldStateIsRemoved
+import com.splendo.kaluga.state.HotStateFlowRepo
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.MainScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.withContext
+import kotlin.time.Duration
+import kotlin.time.TimeSource
+import com.splendo.kaluga.state.State as KalugaState
+
+/**
+ * Timer based on the system clock.
+ * @param duration timer duration
+ * @param interval timer tick interval
+ * @param coroutineScope a parent coroutine scope for the timer
+ */
+class RecurringTimer(
+    override val duration: Duration,
+    interval: Duration,
+    coroutineScope: CoroutineScope = MainScope()
+) : Timer {
+
+    private val stateRepo = TimerStateRepo(duration, interval, coroutineScope)
+    override val state: StateFlow<Timer.State> = stateRepo.stateFlow
+
+    override suspend fun start() = stateRepo.start()
+
+    override suspend fun pause() = stateRepo.pause()
+}
+
+/** Timer state machine. */
+private class TimerStateRepo(
+    totalDuration: Duration,
+    private val interval: Duration,
+    private val coroutineScope: CoroutineScope
+) : HotStateFlowRepo<TimerStateRepo.State>(
+    coroutineScope.coroutineContext,
+    {
+        State.NotRunning.Paused(elapsedSoFar = Duration.ZERO, totalDuration = totalDuration)
+    }
+) {
+    suspend fun start() {
+        withContext(coroutineScope.coroutineContext) {
+            takeAndChangeState { state ->
+                when (state) {
+                    is State.NotRunning.Paused -> suspend {
+                        state.start(interval, coroutineScope, ::finish)
+                    }
+                    is State.NotRunning.Finished, is State.Running -> state.remain()
```

Comment:
when state is Finished shouldn't it switch back to finished?

Reply:
if a start is called in a finished state it shall remain in finished, not restarted

Reply:
Agreed. Its better to have timers not support restarting for now.